js/src/forum/resolver/DiscussionPageResolver.ts

```diff
@@ -0,0 +1,47 @@
+import DefaultResolver from '../../common/resolvers/DefaultResolver';
+
+/**
+ * This isn't exported as it is a temporary measure.
+ * A more robust system will be implemented alongside UTF-8 support in beta 15.
+ */
+function getDiscussionIdFromSlug(slug: string | undefined) {
+  if (!slug) return;
+  return slug.split('-')[0];
+}
+
+/**
+ * A custom route resolver for DiscussionPage that generates the same key to all posts
+ * on the same discussion. It triggers a scroll when going from one post to another
+ * in the same discussion.
+ */
+export default class DiscussionPageResolver extends DefaultResolver {
+  static scrollToPostNumber: number | null = null;
+
+  makeKey() {
+    const params = { ...m.route.param() };
+    if ('near' in params) {
+      delete params.near;
+    }
+    params.id = getDiscussionIdFromSlug(params.id);
+    return this.routeName.replace('.near', '') + JSON.stringify(params);
+  }
+
+  onmatch(args, requestedPath, route) {
+    if (route.includes('/d/:id') && getDiscussionIdFromSlug(args.id) === getDiscussionIdFromSlug(m.route.param('id'))) {
```

Comment:
What would happen if we remove the route includes check? Isn't the `onmatch` of the DiscussionPageResolver only run for discussion routes?

Reply:
So we need to check 2 parts: that we are currently on a DiscussionPage, and that the page we are going to is a DiscussionPage. That we are in `onmatch` of the DiscussionPageResolver gives us the first bit, but we still need to ensure the second. That's why there's no `app.current.matches(DiscussionPage)` here.

Reply:
I see what you mean, but I'm not sure that this check would work.
 
1. From IndexPage to DiscussionPage : DiscussionPageResolver onmatch is run and route includes '/d/:id'
2. From DiscussionPage to DiscussionPage : same, route includes '/d/:id'
3. From DiscussionPage to IndexPage : DiscussionPageResolver onmatch is not run, instead we run the DefaultResolver onmatch.

We can remove the route.includes statement because it would be always true when DiscussionPageResolver is run. 
If we want to check whether we're going from DiscussionPage to DiscussionPage (to avoid the highly unlikely edge case that the previous page has also an id param and the id's match), using `app.current.matches(DiscussionPage)` should work.  

Here's the console.log statement that I used in onmatch to debug this:
```js    
console.log({
      route,
      requestedPath,
      id: getDiscussionIdFromSlug(args.id),
      oldId: getDiscussionIdFromSlug(m.route.param('id')),
      getRoute: m.route.get(),
      matchesAppCurrent: app.current.matches(DiscussionPage),
    });
`

Reply:
Of course... onmatch runs for the resolver of the route we are going to, not of the route we are currently on :facepalming: